lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,936 @@
+
+        self.map_to_cbet_preds_bin: outputs probability of each code (for each layer).
+            The input dimension of `self.map_to_cbet_preds_bin` is same with the output of GPT,
+            and the output dimension of `self.map_to_cbet_preds_bin` is `self.vqvae_model.vqvae_num_layers (=fixed as 2) * self.config.vqvae_n_embed`.
+            if the agent select the code sequentially, we use self.map_to_cbet_preds_primary_bin and self.map_to_cbet_preds_secondary_bin instead of self._map_to_cbet_preds_bin.
+
+        self.map_to_cbet_preds_offset: output the predicted offsets for all the codes in all the layers.
+            The input dimension of ` self.map_to_cbet_preds_offset` is same with the output of GPT,
+            and the output dimension of ` self.map_to_cbet_preds_offset` is `self.vqvae_model.vqvae_num_layers (=fixed as 2) * self.config.vqvae_n_embed * config.action_chunk_size * config.output_shapes["action"][0]`.
+        """
+
+        super().__init__()
+        self.config = config
+        # init vqvae
+        self.vqvae_model = VqVae(config)
+        if config.sequentially_select:
+            self.map_to_cbet_preds_primary_bin = MLP(
+                in_channels=config.gpt_output_dim,
+                hidden_channels=[self.config.vqvae_n_embed],
+            )
+            self.map_to_cbet_preds_secondary_bin = MLP(
+                in_channels=config.gpt_output_dim + self.config.vqvae_n_embed,
+                hidden_channels=[self.config.vqvae_n_embed],
+            )
+        else:
+            self.map_to_cbet_preds_bin = MLP(
+                in_channels=config.gpt_output_dim,
+                hidden_channels=[self.vqvae_model.vqvae_num_layers * self.config.vqvae_n_embed],
+            )
+        self.map_to_cbet_preds_offset = MLP(
+            in_channels=config.gpt_output_dim,
+            hidden_channels=[
+                self.vqvae_model.vqvae_num_layers
+                * self.config.vqvae_n_embed
+                * config.action_chunk_size
+                * config.output_shapes["action"][0],
+            ],
+        )
+        # loss
+        self._focal_loss_fn = FocalLoss(gamma=2.0)
+
+    def discretize(self, n_vqvae_training_steps, actions):
+        if self.vqvae_model.config.action_chunk_size == 1:
+            # even though this code is not using action chunk, it maintains a dummy dimension for action chunk (batch, action_chunk_size = 1, action_dim)
+            actions = actions.reshape(-1, 1, actions.shape[-1])
+        else:
+            # using action chunk
+            slices = []
+            slices.extend(
+                [
+                    actions[:, j : j + self.vqvae_model.config.action_chunk_size, :]
+                    for j in range(actions.shape[1] + 1 - self.vqvae_model.config.action_chunk_size)
+                ]
+            )
+            # dimensions : (batch, action_chunk_size, action_dim)
```

Comment:
I think this comment may be misleading. Shouldn't it be
```suggestion
            # dimensions : (batch * n_chunks, action_chunk_size, action_dim)
```

Reply:
In fact, it might help to explain what is going on here. We have a batch with an action sequence that is longer than the chunk size. We are making use of this by grabbing all possible chunks of actions from the batch. We do this with a sliding window in the list comprehension.

Reply:
Thank you! I updated the comments in this part:


```
    def discretize(self, n_vqvae_training_steps, actions):
        # Resize the action sequence data to fit the action chunk size using a sliding window approach.
        actions = torch.cat(
                [
                    actions[:, j : j + self.config.action_chunk_size, :]
                    for j in range(actions.shape[1] + 1 - self.config.action_chunk_size)
                ],
                dim=0
            )
        # `actions` is a tensor of shape (new_batch, action_chunk_size, action_dim) where new_batch is the number of possible chunks created from the original sequences using the sliding window.
```